### ISSUE TITLE
syntax error at end of input

### ISSUE THREAD

**Issue by VanderneySouza** [jcachat/EntityFramework.DynamicFilters#64]
Hi, how are you?

I have a problem adding more than one filter in context.
When trying to run a query with both filters enabled, it returns 'syntax error at end of input'.

If you use only one of them works normally.

Follows some error data.

**StackTrace:** 
   em System.Data.Entity.Core.EntityClient.Internal.EntityCommandDefinition.ExecuteStoreCommands(EntityCommand entityCommand, CommandBehavior behavior)
   em System.Data.Entity.Core.Objects.Internal.ObjectQueryExecutionPlan.Execute[TResultType](ObjectContext context, ObjectParameterCollection parameterValues)
   em System.Data.Entity.Core.Objects.ObjectQuery`1.<>c__DisplayClass7.<GetResults>b__6()
   em System.Data.Entity.Core.Objects.ObjectContext.ExecuteInTransaction[T](Func`1 func, IDbExecutionStrategy executionStrategy, Boolean startLocalTransaction, Boolean releaseConnectionOnSuccess)
   em System.Data.Entity.Core.Objects.ObjectQuery`1.<>c__DisplayClass7.<GetResults>b__5()
   em System.Data.Entity.Infrastructure.DefaultExecutionStrategy.Execute[TResult](Func`1 operation)
   em System.Data.Entity.Core.Objects.ObjectQuery`1.GetResults(Nullable`1 forMergeOption)
   em System.Data.Entity.Core.Objects.ObjectQuery`1.<System.Collections.Generic.IEnumerable<T>.GetEnumerator>b__0()
   em System.Data.Entity.Internal.LazyEnumerator`1.MoveNext()
   em System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   em System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   em VSoftware.Erp.UI.Program.Main() na c:\Program.cs:linha 28
   em System.AppDomain._nExecuteAssembly(RuntimeAssembly assembly, String[] args)
   em System.AppDomain.nExecuteAssembly(RuntimeAssembly assembly, String[] args)
   em System.Runtime.Hosting.ManifestRunner.Run(Boolean checkAptModel)
   em System.Runtime.Hosting.ManifestRunner.ExecuteAsAssembly()
   em System.Runtime.Hosting.ApplicationActivator.CreateInstance(ActivationContext activationContext, String[] activationCustomData)
   em System.Runtime.Hosting.ApplicationActivator.CreateInstance(ActivationContext activationContext)
   em System.Activator.CreateInstance(ActivationContext activationContext)
   em Microsoft.VisualStudio.HostingProcess.HostProc.RunUsersAssemblyDebugInZone()
   em System.Threading.ThreadHelper.ThreadStart_Context(Object state)
   em System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   em System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   em System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   em System.Threading.ThreadHelper.ThreadStart()

**Filters**:

```
modelBuilder.Filter("FiltroPendenciaAutorizacao", (IBaseGenerica x) => x.PendenteAprovacao, false);
modelBuilder.Filter("FiltroPorEmpresa", (IBaseGenerica x) => x.IdEmpresaCadastro,
                () => ConfiguracaoGeralDominio.RetornaEmpresaLogada().Id);
```

**InternalQuery in EF:**

` SELECT "Var_1"."Id" FROM "config"."Empresas" AS "Var_1" WHERE ("Var_1"."IdEmEmprCada" = @DynamicFilterParam_000001 OR @DynamicFilterParam_000002 IS NOT NULL) AND ("Var_1"."PendenteAprovacao" = @DynamicFilterParam_000003 OR @DynamicFilterParam_000004 IS NOT NULL)`

yours sincerely,

Vanderney Souza

**Comment by jcachat** [comment]
What database are you using?  I think that is a common Postgres error.  Can you capture the full sql (including parameters) that is being sent?  And is there any more detail on the error?

The sql looks valid to me but I am not familiar with Postgres so if that's what you are using, maybe there is something wrong with it that I'm not seeing.

**Comment by VanderneySouza** [comment]
@jcachat  Yes, I'm using Postgres.
How can I capture the SQL in more detail?

The values of the parameters:

@ DynamicFilterParam_000001 is integer - value informing = 1
@ DynamicFilterParam_000003 is boolean - value = false reporting.

As for the parameters @ DynamicFilterParam_000002 and @DynamicFilterParam_000004 I dont know when you are informed because they are assembled by the filter.

Unfortunately on the exceptions is not returned nothing more of information.

yours sincerely,

Vanderney Souza

**Comment by VanderneySouza** [comment]
Hello.
As requested, I am adding more details about the case.

###  **Scenario 1: Using only 1 filter (without applying other conditions) - It works**
_Context.GrupoUsuario();_

**result IQueryble.ToString() -** 
```
SELECT \"Var_1\".\"Id\", \"Var_1\".\"Descricao\", \"Var_1\".\"Ativo\", \"Var_1\".\"IdEmEmprCada\", \"Var_1\".\"PendenteAprovacao\" 
FROM \"config\".\"GruposUsuarios\" AS \"Var_1\" 
WHERE \"Var_1\".\"IdEmEmprCada\" = @DynamicFilterParam_000001 OR @DynamicFilterParam_000002 IS NOT NULL
```

**result Database.Log -**
```
SELECT "Var_1"."Id", "Var_1"."Descricao", "Var_1"."Ativo", "Var_1"."IdEmEmprCada", "Var_1"."PendenteAprovacao" 
FROM "config"."GruposUsuarios" AS "Var_1" 
WHERE "Var_1"."IdEmEmprCada" = @DynamicFilterParam_000001 OR @DynamicFilterParam_000002 IS NOT NULL
```

-- DynamicFilterParam_000001: '0' (Type = Int32, IsNullable = false)
-- DynamicFilterParam_000002: 'null' (Type = Boolean, IsNullable = false)

### **Scenario 2: Using 2 filters (without applying other conditions) - Does not work, apparently the this incomplete command.**
_Context.GrupoUsuario();_

**result IQueryble.ToString() -** 
```
SELECT \"Var_1\".\"Id\", \"Var_1\".\"IdEmEmprCada\", \"Var_1\".\"PendenteAprovacao\", \"Var_1\".\"Descricao\", \"Var_1\".\"Ativo\" 
FROM \"config\".\"GruposUsuarios\" AS \"Var_1\" 
WHERE (\"Var_1\".\"PendenteAprovacao\" = @DynamicFilterParam_000001 OR @DynamicFilterParam_000002 IS NOT NULL) 
AND (\"Var_1\".\"IdEmEmprCada\" = @DynamicFilterParam_000003 OR @DynamicFilterParam_000004 IS NOT NULL)
```

**result Database.Log -**
```
SELECT "Var_1"."Id", "Var_1"."Descricao", "Var_1"."Ativo", "Var_1"."IdEmEmprCada", "Var_1"."PendenteAprovacao" 
FROM "config"."GruposUsuarios" AS "Var_1" 
WHERE ("Var_1"."PendenteAprovacao" = @DynamicFilterParam_000001  
AND ("Var_1"."IdEmEmprCada" = @DynamicFilterParam_000003 
```

-- DynamicFilterParam_000001: '0' (Type = Int32, IsNullable = false)
-- DynamicFilterParam_000002: 'null' (Type = Boolean, IsNullable = false)
-- DynamicFilterParam_000003: '0' (Type = Int32, IsNullable = false)
-- DynamicFilterParam_000004: 'null' (Type = Boolean, IsNullable = false)

### **Scenario 3: Using only 1 filter with other conditions - Does not work because is missing a parenthesis ')'**
_Context.GrupoUsuario.Where(m => m.Descricao == "Administrador");_

**result IQueryble.ToString() -**
```
SELECT \"Extent1\".\"Id\", \"Extent1\".\"Descricao\", \"Extent1\".\"Ativo\", \"Extent1\".\"IdEmEmprCada\", \"Extent1\".\"PendenteAprovacao\" 
FROM \"config\".\"GruposUsuarios\" AS \"Extent1\" 
WHERE (\"Extent1\".\"IdEmEmprCada\" = @DynamicFilterParam_000001 OR @DynamicFilterParam_000002 IS NOT NULL) 
AND (\"Extent1\".\"Descricao\" = @p__linq__0 OR \"Extent1\".\"Descricao\" IS NULL AND @p__linq__0 IS NULL)
```

**result Database.Log -**
```
SELECT "Extent1"."Id", "Extent1"."Descricao", "Extent1"."Ativo", "Extent1"."IdEmEmprCada", "Extent1"."PendenteAprovacao" 
FROM "config"."GruposUsuarios" AS "Extent1" 
WHERE ("Extent1"."IdEmEmprCada" = @DynamicFilterParam_000001  
AND ("Extent1"."Descricao" = @p__linq__0 OR "Extent1"."Descricao" IS NULL AND @p__linq__0 IS NULL)
```

-- DynamicFilterParam_000001: '0' (Type = Int32, IsNullable = false)
-- DynamicFilterParam_000002: 'null' (Type = Boolean, IsNullable = false)
-- p__linq__0: 'Administrador' (Type = Object)

### **Scenario 4: Using 2 filter with other conditions - Does not work, apparently the this incomplete command.**
_Context.GrupoUsuario.Where(m => m.Descricao == "Administrador");_

**result IQueryble.ToString() -**
```
SELECT \"Extent1\".\"Id\", \"Extent1\".\"IdEmEmprCada\", \"Extent1\".\"PendenteAprovacao\", \"Extent1\".\"Descricao\", \"Extent1\".\"Ativo\" 
FROM \"config\".\"GruposUsuarios\" AS \"Extent1\" 
WHERE (\"Extent1\".\"PendenteAprovacao\" = @DynamicFilterParam_000001 OR @DynamicFilterParam_000002 IS NOT NULL) 
AND (\"Extent1\".\"IdEmEmprCada\" = @DynamicFilterParam_000003 OR @DynamicFilterParam_000004 IS NOT NULL) 
AND (\"Extent1\".\"Descricao\" = @p__linq__0 OR \"Extent1\".\"Descricao\" IS NULL AND @p__linq__0 IS NULL)
```

**result Database.Log -**
```
SELECT "Extent1"."Id", "Extent1"."IdEmEmprCada", "Extent1"."PendenteAprovacao", "Extent1"."Descricao", "Extent1"."Ativo" 
FROM "config"."GruposUsuarios" AS "Extent1" 
WHERE ("Extent1"."PendenteAprovacao" = @DynamicFilterParam_000001  
AND ("Extent1"."IdEmEmprCada" = @DynamicFilterParam_000003  
AND ("Extent1"."Descricao" = @p__linq__0 OR "Extent1"."Descricao" IS NULL AND @p__linq__0 IS NULL)
```

-- DynamicFilterParam_000001: '0' (Type = Int32, IsNullable = false)
-- DynamicFilterParam_000002: 'null' (Type = Boolean, IsNullable = false)
-- DynamicFilterParam_000003: '0' (Type = Int32, IsNullable = false)
-- DynamicFilterParam_000004: 'null' (Type = Boolean, IsNullable = false)
-- p__linq__0: 'Administrador' (Type = Object)

yours sincerely,

Vanderney Souza

**Comment by leddt** [comment]
I believe this is caused by the changes related to #53. More specifically `RemoveFilterDisabledConditionFromQuery`. It will need to be smarter about how it removes the clause from the string.

**Comment by jcachat** [comment]
Yes, I think you're right.  Can you try downgrading to 1.4.9 of the library and see if everything works?  That is the build before any of the changes in #53 were introduced.

It's odd though that all of the unit tests pass.  So it may have something to do with Postgres specific syntax.  That's the danger of fiddling with the sql text like this...  I will need to get Postgres installed (which I've been meaning to do for a long time) and specifically test against that database.

But I'm on vacation atm and won't be back for another week.  And I'm not equipped to work on this from here.  So I won't be able to look into this more until late next week.

**Comment by VanderneySouza** [comment]
I managed to find the problem in relation to Postgree.
As mentioning by @leddt , I did a quick analysis on the function  

`RemoveFilterDisabledConditionFromQuery` and on the line,

`command.CommandText = command.CommandText.Remove (startIdx, endIdx - startIdx + 1);`

The command was removing the parenthesis that closes the expression.
In this case, I removed +1 and worked the Postgree;

`command.CommandText = command.CommandText.Remove (startIdx, endIdx - startIdx);`

I have to analyze to better understand how it works, because this was my first contact with her, and I do not know exactly how DLL it works.

yours sincerely,

Vanderney Souza

**Comment by jcachat** [comment]
I just pushed out a fix and updated the Nuget package.

The problem was that PostgreSQL generates the SQL a little differently than the other DBs I tested against:

PostgreSQL:
```
("Var_1"."ID" < @DynamicFilterParam_000001 OR @DynamicFilterParam_000002 IS NOT NULL)
```
Others:
```
(("Var_1"."ID" < @DynamicFilterParam_000001) OR (@DynamicFilterParam_000002 IS NOT NULL))
```

So RemoveFilterDisabledConditionFromQuery() was removing the trailing ')' when it should not have.  Your fix was correct in most cases, but if there is a single condition, there would be no trailing ')' at all.  It would actually generate SQL like this:
```
"Var_1"."ID" < @DynamicFilterParam_000001 OR @DynamicFilterParam_000002 IS NOT NULL
```

I (finally) got PostgreSQL installed on my machine so I was able to run the full set of unit tests.  So make sure you get this latest update for the full fix.
